Added new profile mapping configuration for Teradata

sc250072 opened the pull request:
## Description

<!-- Add a brief but complete description of the change. -->
Teradata has [Provider](https://airflow.apache.org/docs/apache-airflow-providers-teradata/stable/index.html) in airflow and [adapter](https://github.com/Teradata/dbt-teradata)  in dbt. The  cosmos library doesn't have profile configuration with mapping support. This PR address this issue.

## Related Issue(s)

<!-- If this PR closes an issue, you can use a keyword to auto-close. -->
<!-- i.e. "closes #0000" -->
Closes https://github.com/astronomer/astronomer-cosmos/issues/1053

## Breaking Change?

<!-- If this introduces a breaking change, specify that here. -->
No Breaking Change.

## Checklist

- [ ] I have made corresponding changes to the documentation (if required)
- [x] I have added tests that prove my fix is effective or that my feature works

tatiana commented on cosmos/profiles/teradata/user_pass.py:
```diff
@@ -0,0 +1,62 @@
+
+        return self.filter_null(profile)
+
+    @property
+    def mock_profile(self) -> dict[str, Any | None]:
+        """Gets mock profile."""
+        parent_mock = super().mock_profile
+
+        return {
+            **parent_mock,
+        }
```
Do we need to implement this method, or can we rely on the default one by inheritance?

sc250072 replied:
Yes agree. Nothing special handling than inherited mock_profile. Removed it.

tatiana commented on cosmos/profiles/teradata/user_pass.py:
```diff
@@ -0,0 +1,62 @@
+    """
+
+    airflow_connection_type: str = "teradata"
+    dbt_profile_type: str = "teradata"
```
Could you also add `is_community`, setting it to true?

```suggestion
    dbt_profile_type: str = "teradata"
    is_community: bool = True

```

tatiana replied:
@sc250072 is important because Astronomer does not have credentials to Teradata which would allow us to validate these changes.So we are relying on community members to implement/validate and maintain this `ProfileMapping`.

tatiana reviewed (commented):
This looks great, @sc250072 , thank you very much for adding support to Teradata! I added some minor comments in-line

tatiana commented on cosmos/profiles/teradata/user_pass.py:
```diff
@@ -0,0 +1,62 @@
+
+from ..base import BaseProfileMapping
+
+if TYPE_CHECKING:
```
Why do we have this?

sc250072 replied:
Removed it. Its not required.

sc250072 commented:
> This looks great, @sc250072 , thank you very much for adding support to Teradata! I added some minor comments in-line

Thank you @tatiana . Addressed the review comments.

tatiana commented on tests/profiles/teradata/test_teradata_user_pass.py:
```diff
@@ -0,0 +1,176 @@
+        login="my_user",
+        password="my_password",
+        schema="my_database",
+        extra='{"tmode": "TERA"}',
```
At the moment (2 July 2024), it seems dbt only supports the "ANSI" mode:
https://docs.getdbt.com/docs/core/connect-data-platform/teradata-setup

Since this is a unit test, it probably does not matter:
![Screenshot 2024-07-02 at 23 02 20](https://github.com/astronomer/astronomer-cosmos/assets/272048/f0c941ad-fece-4cbf-be49-88e6886e1fa2)

sc250072 replied:
@tatiana recently, we added support of TERA mode as well https://github.com/Teradata/dbt-teradata/blob/main/dbt/adapters/teradata/connections.py. Will check with our teradata dbt team and update this doc accordingly.

tatiana replied:
Thanks a lot, @sc250072 , I'll resolve this conversation - from a Cosmos perspective it's a non-isssue

tatiana reviewed (approved):
Thanks, @sc250072 for being so quick at addressing the comments. Happy for us to merge this PR once all the checks pass

sc250072 commented:
> Thanks, @sc250072 for being so quick at addressing the comments. Happy for us to merge this PR once all the checks pass

Thank you for the quick response on the PR. Right now, I'm conducting integration testing by manually installing the .whl file on my local machine. Could you please provide any documentation or guidance on implementing integration tests that can be run through GitHub workflows?

pankajastro commented on cosmos/profiles/teradata/user_pass.py:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @property
+    def profile(self) -> dict[str, Any | None]:
```
```suggestion
    def profile(self) -> dict[str, Any]:
```
why none?

pankajastro commented:
We have test file https://github.com/astronomer/astronomer-cosmos/blob/main/tests/test_example_dags.py which run some dags of this repo https://github.com/astronomer/astronomer-cosmos/tree/main/dev/dags also you can find some tests decorated with `@pytest.mark.integration` runs as integration tests

pankajastro reviewed (approved):
lgtm

tatiana commented:
> We have test file https://github.com/astronomer/astronomer-cosmos/blob/main/tests/test_example_dags.py which run some dags of this repo https://github.com/astronomer/astronomer-cosmos/tree/main/dev/dags also you can find some tests decorated with `@pytest.mark.integration` runs as integration tests. But running integration tests from this repo may require some Teradata credentials so I'm not sure if you have those at the moment

@sc250072 complemeting @pankajastro 's message: https://github.com/astronomer/astronomer-cosmos/pull/1077#issuecomment-2205271408

Could the Teradata team share specific credentials to run integration tests from the Cosmos repo to validate that these tests pass?  In that case, we'd be happy to remove the community flag.

As @pankajastro mentioned, we could either have an example DAG with Teradata or other tests marked with integration.

Another alternative, similar to Airflow, is you run a Cosmos DAG using this profile in your own CI/CD against the Cosmos main branch and share a dashboard/page with the job status with us. In this case, you wouldn't have to share credentials with us, and we could help guarantee the stability of this feature. In this case, we could remove the `is_community` flag again.

tatiana commented:
@sc250072 this work looks great, we'll be merging this PR and we can address removing the `is_community` in a separate PR

sc250072 commented:
> @sc250072 this work looks great, we'll be merging this PR and we can address removing the `is_community` in a separate PR

Thank you @tatiana .

sc250072 commented:
@pankajastro @tatiana any expected date of new version of astronomer-cosmos, which includes teradata support.  https://github.com/astronomer/astronomer-cosmos/blob/main/CHANGELOG.rst

tatiana commented:
@sc250072 it'll be released as part of [1.6](https://github.com/astronomer/astronomer-cosmos/milestone/20), which is currently planned for the end of July.

Would you like us to create an alpha (e.g., 1.6.0a1) for you to validate the feature beforehand?

sc250072 commented:
> @sc250072 it'll be released as part of [1.6](https://github.com/astronomer/astronomer-cosmos/milestone/20), which is currently planned for the end of July.
> 
> Would you like us to create an alpha (e.g., 1.6.0a1) for you to validate the feature beforehand?

Yes that would be help to validate feature before release.

tatiana commented:
@sc250072 you can validate the feature in Cosmos 1.6.0a1
* https://pypi.org/project/astronomer-cosmos/1.6.0a1/
* https://github.com/astronomer/astronomer-cosmos/releases/tag/astronomer-cosmos-v1.6.0a1

sc250072 commented:
@tatiana  I have tested Cosmos 1.6.0a1 and identified an issue with the Teradata mock profile, documented here: https://github.com/astronomer/astronomer-cosmos/issues/1087. I have resolved this issue and submitted a pull request at https://github.com/astronomer/astronomer-cosmos/pull/1088.

After addressing the issue, I validated the fix using a locally built .whl file and tested it with several Cosmos DAGs within the dbt-jaffle-shop project, utilizing the dbt-teradata adapter. Everything is functioning correctly now.

Could we proceed with another alpha release incorporating these changes?"

tatiana commented:
Hi @sc250072 , thanks for all the work on this!

Yes, we'll be releasing a new alpha this week Nd we'll let you know

sc250072 commented:
> Hi @sc250072 , thanks for all the work on this!
> 
> Yes, we'll be releasing a new alpha this week Nd we'll let you know

@tatiana any update on new alpha release
